cpp/src/neighbors/detail/vamana/vamana_build.cuh

```diff
@@ -0,0 +1,471 @@
+      GreedySearchKernel<T, accT, IdxT, Accessor>
+          <<<num_blocks, blockD, search_smem_total_size>>>(
+                     d_graph.view(), 
+                     dataset, 
+                     query_list_ptr.data(), 
+                     step_size, 
+                     medoid_id,
+                     degree, 
+                     dataset.extent(0), 
+                     visited_size,
+                     metric, 
+		     dim,
+		     queue_size,
+		     search_smem_sort_size);
+
+
+        // Run on candidates of vectors being inserted
+      RobustPruneKernel<T,accT, IdxT>
+          <<<num_blocks, blockD, prune_smem_total_size>>>(
+                     d_graph.view(), 
+                     dataset,
+                     query_list_ptr.data(),
+                      step_size,
+                     degree,
+		     visited_size,
+                     dataset.extent(0),
+                     metric,
+                     alpha,
+		     dim,
+		     prune_smem_sort_size);
+		     
+
+// Write results from first prune to graph edge list
+      write_graph_edges_kernel<accT, IdxT><<<num_blocks, blockD>>>(
+                     d_graph.view(), 
+                     query_list_ptr.data(), 
+                     degree, 
+                     step_size);
+
+// compute prefix sums of query_list sizes - TODO parallelize prefix sums
+      rmm::device_uvector<int> d_total_edges(
+          1, stream, raft::resource::get_large_workspace_resource(res));
+      prefix_sums_sizes<accT,IdxT><<<1,1>>>(query_list, step_size, d_total_edges.data());
+//      cudaDeviceSynchronize(); // TODO -remove?
+
+      int total_edges;
+      raft::copy(&total_edges, d_total_edges.data(), 1, stream);
+  
+// Create reverse edge list
+      create_reverse_edge_list<accT,IdxT><<<num_blocks,blockD>>>(
+                     query_list_ptr.data(), 
+                     step_size, 
+                     degree, 
+                     edge_src.data(), 
+                     edge_dest.data());
+
+      // Sort to group reverse edges by destination
+      cub::DeviceMergeSort::SortPairs(
+                     temp_sort_storage.data(),
+                     temp_storage_bytes,
+                     edge_dest.data(),
+                     edge_src.data(),
+                     total_edges,
+                     CmpEdge<IdxT>());
+
+// Get number of unique node destinations
+     IdxT unique_dests = raft::sparse::neighbors::get_n_components(edge_dest.data(), total_edges, stream);
+
+     // Find which node IDs have reverse edges and their indices in the reverse edge list
+     thrust::device_vector<IdxT> edge_dest_vec(edge_dest.data(), edge_dest.data()+total_edges);
+     thrust::device_vector<int> unique_indices(total_edges);
+     thrust::sequence(unique_indices.begin(), unique_indices.end());
+     thrust::unique_by_key(edge_dest_vec.begin(), edge_dest_vec.end(), unique_indices.begin());
```

Comment:
I haven't been able to find a raft equivalent of `thrust::unique_by_key`. Is it preferable for me to write a custom kernel to do this over using thrust?

Reply:
I think using thrust is fine then. The less code the better :)

Reply:
Ok cool. Moved everything out of using `thrust` where possible, then, but still call `unique_by_key` here.

Reply:
I think this is a `segmented unique`. We should consider abstracting the thrust call behind a RAFT function so that we can centralize it (and update, change, optimize, reuse) it when needed. In general, we've been trying to reduce the direct use of Thrust. I also understand that this is an important feature and we shouldn't hold it up over this, but we could at least create an issue for this and reference the issue in a comment here in the code so we don't lose sight of it.